### PR TITLE
MF-552 - Use event sourcing to keep Bootstrap service in sync with Things service

### LINKED ISSUE #552: Use event sourcing to keep Bootstrap service in sync with Things service
**FEATURE REQUEST**

1. Is there an open issue addressing this request? If it does, please add a "+1" reaction to the
   existing issue, otherwise proceed to step 2.
No.

2. Describe the feature you are requesting, as well as the possible use case(s) for it.
When we close #551, Bootstrap Channels should be kept in sync with Things Channels. Bootstrap service data should be updated in case of:
    - Channel removal
    - Channel update
    - Thing removal
    - Thing update

    Connecting/disconnecting corresponding Mainflux Thing to/from Channel should also be considered.

3. Indicate the importance of this feature to you (must-have, should-have, nice-to-have).
This is a must-have feature.

### PR DESCRIPTION
**Pull request by dborovcanin**
### What does this do?
This PR adds the use of event sourcing from Things service to keep corresponding Bootstrap service entities in sync.

### Which issue(s) does this PR fix/relate to?
This PR resolves #552.

### List any changes that modify/break current functionality
There are 2 major API changes:
    - List of Configs does not return Channels
    - Update Config is split to Config fields update and Config connections update

### Have you included tests for your changes?
Yes.

### Did you document any new/modified functionality?
Yes, docs are up to date.

### CODE REVIEW

**Review by drasko — DISMISSED** [review]
LGTM

**Review by manuio — DISMISSED** [review]

**Review by drasko — DISMISSED** [review]
LGTM

**Review by manuio — DISMISSED** [review]

File: bootstrap/api/endpoint_test.go [hunk 1]
```diff
@@ -351,10 +359,17 @@ func TestView(t *testing.T) {
 		assert.Nil(t, err, fmt.Sprintf("%s: unexpected error %s", tc.desc, err))
 
 		assert.Equal(t, tc.status, res.StatusCode, fmt.Sprintf("%s: expected status code %d got %d", tc.desc, tc.status, res.StatusCode))
-		body, err := ioutil.ReadAll(res.Body)
-		assert.Nil(t, err, fmt.Sprintf("%s: unexpected error %s", tc.desc, err))
-		data := strings.Trim(string(body), "\n")
-		assert.Equal(t, tc.res, data, fmt.Sprintf("%s: expected response '%s' got '%s'", tc.desc, tc.res, data))
+		var view config
+		err = json.NewDecoder(res.Body).Decode(&view)
+		if err != io.EOF {
```
Comment (anovakovic01) [inline bootstrap/api/endpoint_test.go]: Inline.

File: bootstrap/api/logging.go [hunk 2]
```diff
@@ -119,3 +132,55 @@ func (lm *loggingMiddleware) ChangeState(key, id string, state bootstrap.State)
...
+		}
+		lm.logger.Info(fmt.Sprintf("%s without errors.", message))
+	}(time.Now())
+
+	return lm.svc.UpdateChannelHandler(channel)
+}
+
+func (lm *loggingMiddleware) RemoveConfigHandler(id string) (err error) {
+	defer func(begin time.Time) {
+		message := fmt.Sprintf("Method removeConfig for config %s took %s to complete", id, time.Since(begin))
```
Comment (anovakovic01) [inline bootstrap/api/logging.go]: Same here.

File: bootstrap/api/logging.go [hunk 3]
```diff
@@ -119,3 +132,55 @@ func (lm *loggingMiddleware) ChangeState(key, id string, state bootstrap.State)
 
 	return lm.svc.ChangeState(key, id, state)
 }
+
+func (lm *loggingMiddleware) UpdateChannelHandler(channel bootstrap.Channel) (err error) {
+	defer func(begin time.Time) {
+		message := fmt.Sprintf("Method updateChannel for channel %s took %s to complete", channel.ID, time.Since(begin))
```
Comment (anovakovic01) [inline bootstrap/api/logging.go]: Same here.

File: bootstrap/postgres/configs.go [hunk 4]
```diff
@@ -32,68 +36,97 @@ type configRepository struct {
...
+	rows, err := cr.db.Query(q, cfg.MFThing, cfg.Owner)
+	if err != nil {
+		cr.log.Error(fmt.Sprintf("Failed to retrieve connected due to %s", err))
 		return bootstrap.Config{}, err
 	}
+	defer rows.Close()
+
+	for rows.Next() {
+		c := bootstrap.Channel{}
+		if err = rows.Scan(&c.ID, &c.Name, &c.Metadata); err != nil {
```
Comment (anovakovic01) [inline bootstrap/postgres/configs.go]: Use `:=`.

File: bootstrap/postgres/configs.go [hunk 5]
```diff
@@ -154,49 +181,54 @@ func (cr configRepository) RetrieveAll(key string, filter bootstrap.Filter, offs
...
```
Comment (anovakovic01) [inline bootstrap/postgres/configs.go]: Use `:=`.

File: bootstrap/postgres/configs.go [hunk 6]
```diff
@@ -213,9 +245,45 @@ func (cr configRepository) Update(cfg bootstrap.Config) error {
 	return nil
 }
 
+func (cr configRepository) UpdateConnections(key, id string, channels []bootstrap.Channel, connections []string) error {
+	tx, err := cr.db.Begin()
+	if err != nil {
+		return err
+	}
+
+	if err = insertChannels(key, channels, tx); err != nil {
```
Comment (anovakovic01) [inline bootstrap/postgres/configs.go]: Use `:=`.

File: bootstrap/postgres/configs.go [hunk 7]
```diff
@@ -213,9 +245,45 @@ func (cr configRepository) Update(cfg bootstrap.Config) error {
...
+		cr.rollback("Failed to insert Channels during the update", tx, err)
+
+		return err
+	}
+
+	if err = updateConnections(key, id, connections, tx); err != nil {
```
Comment (anovakovic01) [inline bootstrap/postgres/configs.go]: Use `:=`.

File: bootstrap/postgres/configs.go [hunk 8]
```diff
@@ -240,6 +308,28 @@ func (cr configRepository) ChangeState(key, id string, state bootstrap.State) er
...
+
+	rows, err := cr.db.Query(q, key, pq.Array(ids))
+	if err != nil {
+		return []bootstrap.Channel{}, err
+	}
+
+	var channels []bootstrap.Channel
+	for rows.Next() {
+		var ch bootstrap.Channel
+		if err = rows.Scan(&ch.ID, &ch.Name, &ch.Metadata); err != nil {
```
Comment (anovakovic01) [inline bootstrap/postgres/configs.go]: Use `:=`.

**Review by anovakovic01 — CHANGES_REQUESTED** [review]

**Review by drasko — CHANGES_REQUESTED** [review]

**Review by anovakovic01 — APPROVED** [review]
LGTM!

**Review by drasko — APPROVED** [review]
LGTM

**Pull request merged**